### PR TITLE
Add ASA ShowVersion

### PR DESCRIPTION
**Pull request by ajschroeder**
## Description
Added a parser to process the output from "show version" on an ASA. 

## Motivation and Context
I needed to work with ASAs as part of a project at work. I noticed that there was not a parser for "show version" on an ASA, so I decided to write one.

## Impact (If any)
No negative impact expected

## Screenshots:
2021-06-11T08:54:19: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %AETEST-INFO: |                   Starting testcase SuperFileBasedTesting                    |
2021-06-11T08:54:19: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %AETEST-INFO: |                            Starting section setup                            |
2021-06-11T08:54:19: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %AETEST-INFO: The result of section setup is => PASSED
2021-06-11T08:54:19: %AETEST-INFO: The result of testcase SuperFileBasedTesting is => PASSED
2021-06-11T08:54:19: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %AETEST-INFO: |                            Starting testcase asa                             |
2021-06-11T08:54:19: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %AETEST-INFO: |                            Starting section setup                            |
2021-06-11T08:54:19: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %AETEST-INFO: The result of section setup is => PASSED
2021-06-11T08:54:19: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %AETEST-INFO: |                         Starting section ShowVersion                         |
2021-06-11T08:54:19: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %AETEST-INFO: +..............................................................................+
2021-06-11T08:54:19: %AETEST-INFO: :                     Starting STEP 1: asa -> ShowVersion                      :
2021-06-11T08:54:19: %AETEST-INFO: +..............................................................................+
2021-06-11T08:54:19: %AETEST-INFO: +..............................................................................+
2021-06-11T08:54:19: %AETEST-INFO: :             Starting STEP 1.1: Test Golden -> asa -> ShowVersion             :
2021-06-11T08:54:19: %AETEST-INFO: +..............................................................................+
2021-06-11T08:54:19: %AETEST-INFO: +..............................................................................+
2021-06-11T08:54:19: %AETEST-INFO: :       Starting STEP 1.1.1: Gold -> asa -> ShowVersion -> golden_output       :
2021-06-11T08:54:19: %AETEST-INFO: +..............................................................................+
2021-06-11T08:54:19: %AETEST-INFO: The result of STEP 1.1.1: Gold -> asa -> ShowVersion -> golden_output is => PASSED
2021-06-11T08:54:19: %AETEST-INFO: The result of STEP 1.1: Test Golden -> asa -> ShowVersion is => PASSED
2021-06-11T08:54:19: %AETEST-INFO: +..............................................................................+
2021-06-11T08:54:19: %AETEST-INFO: :             Starting STEP 1.2: Test Empty -> asa -> ShowVersion              :
2021-06-11T08:54:19: %AETEST-INFO: +..............................................................................+
2021-06-11T08:54:19: %AETEST-INFO: +..............................................................................+
2021-06-11T08:54:19: %AETEST-INFO: :          Starting STEP 1.2.1: Empty -> asa -> ShowVersion -> empty           :
2021-06-11T08:54:19: %AETEST-INFO: +..............................................................................+
2021-06-11T08:54:19: %AETEST-INFO: The result of STEP 1.2.1: Empty -> asa -> ShowVersion -> empty is => PASSED
2021-06-11T08:54:19: %AETEST-INFO: The result of STEP 1.2: Test Empty -> asa -> ShowVersion is => PASSED
2021-06-11T08:54:19: %AETEST-INFO: The result of STEP 1: asa -> ShowVersion is => PASSED
2021-06-11T08:54:19: %AETEST-INFO: +----------------------------------------------------------+
2021-06-11T08:54:19: %AETEST-INFO: |                       STEPS Report                       |
2021-06-11T08:54:19: %AETEST-INFO: +----------------------------------------------------------+
2021-06-11T08:54:19: %AETEST-INFO: STEP 1 - asa -> ShowVersion                       Passed
2021-06-11T08:54:19: %AETEST-INFO: STEP 1.1 - Test Golden -> asa -> ShowVersion      Passed
2021-06-11T08:54:19: %AETEST-INFO: STEP 1.1.1 - Gold -> asa -> ShowVersion -> golden_outputPassed
2021-06-11T08:54:19: %AETEST-INFO: STEP 1.2 - Test Empty -> asa -> ShowVersion       Passed
2021-06-11T08:54:19: %AETEST-INFO: STEP 1.2.1 - Empty -> asa -> ShowVersion -> empty Passed
2021-06-11T08:54:19: %AETEST-INFO: ------------------------------------------------------------
2021-06-11T08:54:19: %AETEST-INFO: The result of section ShowVersion is => PASSED
2021-06-11T08:54:19: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %AETEST-INFO: |                           Starting section cleanup                           |
2021-06-11T08:54:19: %AETEST-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %AETEST-INFO: The result of section cleanup is => PASSED
2021-06-11T08:54:19: %AETEST-INFO: The result of testcase asa is => PASSED
2021-06-11T08:54:19: %GENIE-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %GENIE-INFO: |                               Unittest results                               |
2021-06-11T08:54:19: %GENIE-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %GENIE-INFO:  SECTIONS/TESTCASES                                                      RESULT
2021-06-11T08:54:19: %GENIE-INFO: --------------------------------------------------------------------------------
2021-06-11T08:54:19: %GENIE-INFO: .
2021-06-11T08:54:19: %GENIE-INFO: |-- SuperFileBasedTesting                                                 PASSED
2021-06-11T08:54:19: %GENIE-INFO: |   `-- setup                                                             PASSED
2021-06-11T08:54:19: %GENIE-INFO: `-- asa                                                                   PASSED
2021-06-11T08:54:19: %GENIE-INFO:     |-- setup                                                             PASSED
2021-06-11T08:54:19: %GENIE-INFO:     |-- ShowVersion                                                       PASSED
2021-06-11T08:54:19: %GENIE-INFO:     |   |-- Step 1: asa -> ShowVersion                                    PASSED
2021-06-11T08:54:19: %GENIE-INFO:     |   |-- Step 1.1: Test Golden -> asa -> ShowVersion                   PASSED
2021-06-11T08:54:19: %GENIE-INFO:     |   |-- Step 1.1.1: Gold -> asa -> ShowVersion -> golden_output       PASSED
2021-06-11T08:54:19: %GENIE-INFO:     |   |-- Step 1.2: Test Empty -> asa -> ShowVersion                    PASSED
2021-06-11T08:54:19: %GENIE-INFO:     |   `-- Step 1.2.1: Empty -> asa -> ShowVersion -> empty              PASSED
2021-06-11T08:54:19: %GENIE-INFO:     `-- cleanup                                                           PASSED
2021-06-11T08:54:19: %GENIE-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %GENIE-INFO: |                                   Summary                                    |
2021-06-11T08:54:19: %GENIE-INFO: +------------------------------------------------------------------------------+
2021-06-11T08:54:19: %GENIE-INFO:  Number of ABORTED                                                            0
2021-06-11T08:54:19: %GENIE-INFO:  Number of BLOCKED                                                            0
2021-06-11T08:54:19: %GENIE-INFO:  Number of ERRORED                                                            0
2021-06-11T08:54:19: %GENIE-INFO:  Number of FAILED                                                             0
2021-06-11T08:54:19: %GENIE-INFO:  Number of PASSED                                                             2
2021-06-11T08:54:19: %GENIE-INFO:  Number of PASSX                                                              0
2021-06-11T08:54:19: %GENIE-INFO:  Number of SKIPPED                                                            0
2021-06-11T08:54:19: %GENIE-INFO:  Total Number                                                                 2
2021-06-11T08:54:19: %GENIE-INFO:  Success Rate                                                            100.0%
2021-06-11T08:54:19: %GENIE-INFO: --------------------------------------------------------------------------------
2021-06-11T08:54:19: %GENIE-INFO:  Total Passing Unittests                                                      2
2021-06-11T08:54:19: %GENIE-INFO:  Total Failed Unittests                                                       0
2021-06-11T08:54:19: %GENIE-INFO:  Total Errored Unittests                                                      0
2021-06-11T08:54:19: %GENIE-INFO:  Total Unittests                                                              2
2021-06-11T08:54:19: %GENIE-INFO: --------------------------------------------------------------------------------
## Checklist:

- [x] I have updated the changelog.
- [ ] I have updated the documentation (If applicable).
- [ ] I have added tests to cover my changes (If applicable).
- [x] All new and existing tests passed.
- [x] All new code passed compilation.

### CODE REVIEW

File: changelog/2021/JUNE.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+--------------------------------------------------------------------------------
+                                      New                                       
+--------------------------------------------------------------------------------
+
+* asa
+        * Added ShowVersion
```
Comment (LukasMcClelland) [inline changelog/2021/JUNE.rst:6]: Please change to 
```
--------------------------------------------------------------------------------
                                      New 
--------------------------------------------------------------------------------
* asa
    * Added ShowVersion:
      * show version
```

**Review by LukasMcClelland — COMMENTED** [review]
Thank you very much for the contribution! Excellent work and attention to detail. 

Looking through Cisco documentation, I found this example of more device output from the ASA `show version` command. Would you be interested in adapting this parser to be able to handle the output from the below link? It would be great to have a wider range of output supported by this parser. (Please also create test files for the new device output if so)
https://www.cisco.com/c/en/us/td/docs/security/asa/asa-cli-reference/S/asa-command-ref-S/m_show_u-show_z.html#wp3989305387

Please let me know what you decide to do

**Comment by ajschroeder** [comment]
I could try and adapt the parser for the output in the link you sent over. I agree that more support is better, although I don’t know how much ASA version 8.x code is out in the wild these days. Is there a way to test this output without having a device? I did all my testing against an actual ASAv.

From: Lukas McClelland ***@***.***>
Date: Monday, June 14, 2021 at 10:54 AM
To: CiscoTestAutomation/genieparser ***@***.***>
Cc: TheHitman1977 ***@***.***>, Author ***@***.***>
Subject: Re: [CiscoTestAutomation/genieparser] Add ASA ShowVersion (#468)

@LukasMcClelland commented on this pull request.

Thank you very much for the contribution! Excellent work and attention to detail.

Looking through Cisco documentation, I found this example of more device output from the ASA show version command. Would you be interested in adapting this parser to be able to handle the output from the below link? It would be great to have a wider range of output supported by this parser. (Please also create test files for the new device output if so)
https://www.cisco.com/c/en/us/td/docs/security/asa/asa-cli-reference/S/asa-command-ref-S/m_show_u-show_z.html#wp3989305387<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.cisco.com%2Fc%2Fen%2Fus%2Ftd%2Fdocs%2Fsecurity%2Fasa%2Fasa-cli-reference%2FS%2Fasa-command-ref-S%2Fm_show_u-show_z.html%23wp3989305387&data=04%7C01%7C%7C5e04b6da4c534f6d388708d92f4cb683%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637592828782847855%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=pDMAo1vjaDr87gB7iac%2B3WvO1ExS7%2BLDMQV%2BM3t2C3w%3D&reserved=0>

Please let me know what you decide to do

________________________________

In changelog/2021/JUNE.rst<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCiscoTestAutomation%2Fgenieparser%2Fpull%2F468%23discussion_r651049674&data=04%7C01%7C%7C5e04b6da4c534f6d388708d92f4cb683%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637592828782847855%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=2vMSEZ6APt8UJ606pw1lX1atINJX2r0kamg9znkKIOQ%3D&reserved=0>:

> +--------------------------------------------------------------------------------

+                                      New

+--------------------------------------------------------------------------------

+

+* asa

+        * Added ShowVersion

Please change to

--------------------------------------------------------------------------------

                                      New

--------------------------------------------------------------------------------

* asa

    * Added ShowVersion:

      * show version

**Comment by LukasMcClelland** [comment]
Sounds great! You can post any questions/challenges here and I'll check in periodically and help out.

> Is there a way to test this output without having a device?

Easiest way is to make another test with the following files:  
```
genieparser/src/genie/libs/parser/asa/tests/ShowVersion/cli/equal/golden_output1_expected.py
genieparser/src/genie/libs/parser/asa/tests/ShowVersion/cli/equal/golden_output1_output.txt
```
then paste the other device output (from Cisco.com) into `golden_output1_output.txt` and put an empty dictionary into `golden_output1_expected.py`:
```
expected_output = {

}
``` 
You can test the parser by running the following command **from within** `genieparser/tests`:
```
python folder_parsing_job.py -o asa -c ShowVersion -f
```
This tool will show you difference between parsed output and expected output. Use it to help shape your code to accept both device outputs. Thank you very much :)

**Comment by ajschroeder** [comment]
Hi Lucas,

First, thank you for pointing me in the right direction for writing the parser without an actual device. That is great!

So I have what I think is about 90% of both parsers done. Well, the first one was done but I tweaked my schema a bit where I felt it made sense to. The second parser is almost done, I have some other output that I need to account for that is not present in the other ASA. I have it to the point where both tests are passing. Hopefully I can get the rest of the output done in the next week.

Thanks again for the help – I’ll update you when I’m ready with the other parser. I’ll have to brush up on how to re-submit my PR with the updated code.

AJ

From: Lukas McClelland ***@***.***>
Date: Monday, June 14, 2021 at 2:12 PM
To: CiscoTestAutomation/genieparser ***@***.***>
Cc: TheHitman1977 ***@***.***>, Author ***@***.***>
Subject: Re: [CiscoTestAutomation/genieparser] Add ASA ShowVersion (#468)

Sounds great! You can post any questions/challenges here and I'll check in periodically and help out.

Is there a way to test this output without having a device?

Easiest way is to make another test with the following files:

genieparser/src/genie/libs/parser/asa/tests/ShowVersion/cli/equal/golden_output1_expected.py

genieparser/src/genie/libs/parser/asa/tests/ShowVersion/cli/equal/golden_output1_output.txt

then paste the other device output (from Cisco.com) into golden_output1_output.txt and put an empty dictionary into golden_output1_expected.py:

expected_output = {



}

You can test the parser by running the following command from within genieparser/tests:

python folder_parsing_job.py -o asa -c ShowVersion -f

This tool will show you difference between parsed output and expected output. Use it to help shape your code to accept both device outputs. Thank you very much :)

File: changelog/2021/JUNE.rst (hunk 1 above)
Comment (ajschroeder) [reply changelog/2021/JUNE.rst]: This has been changed

**Comment by LukasMcClelland** [comment]
Sounds great! Looking forward to reviewing your contribution!

**Comment by ajschroeder** [comment]
Hi Lukas,

I think I have things to the point of being “done”. The tests pass and I think I have checked in, committed, and pushed my changes to the PR so you should be able to see them at your earliest convenience.

Thanks,
AJ

File: src/genie/libs/parser/asa/show_version.py [hunk 2]
```diff
@@ -0,0 +1,730 @@
...
+        # 7: Ext: GigabitEthernet0/6  : address is 5001.0003.0007, irq 10
+        p11 = re.compile(r'^\s*(?P<intf_number>\d+): '
+                         r'Ext: (?P<interface>[\w\/\.\-]+)\s*: address is '
+                         r'(?P<mac_addr>[0-9A-Fa-f]{4}\.[0-9A-Fa-f]{4}\.[0-9A-Fa-f]{4}), '
+                         r'irq (?P<intf_irq>\d{1,2})')
+
+        # 5: Int: Not used            : irq 11
+        # 6: Int: Not used            : irq 5
+        p11_1 = re.compile(r'^\s*(?P<intf_number>\d+): Int: (?P<interface>[\w ]+)\s*: '
+                           r'irq (?P<intf_irq>\d{1,2})')
```
Comment (LukasMcClelland) [inline src/genie/libs/parser/asa/show_version.py:170]: Can `irq` be followed by a 3 digit number? Perhaps safer to do `\d+`

File: src/genie/libs/parser/asa/show_version.py [hunk 3]
```diff
@@ -0,0 +1,730 @@
...
+        # 1: Ext: GigabitEthernet0/0  : address is 5001.0003.0001, irq 11
+        # 2: Ext: GigabitEthernet0/1  : address is 5001.0003.0002, irq 10
+        # 3: Ext: GigabitEthernet0/2  : address is 5001.0003.0003, irq 10
+        # 4: Ext: GigabitEthernet0/3  : address is 5001.0003.0004, irq 11
+        # 5: Ext: GigabitEthernet0/4  : address is 5001.0003.0005, irq 11
+        # 6: Ext: GigabitEthernet0/5  : address is 5001.0003.0006, irq 10
+        # 7: Ext: GigabitEthernet0/6  : address is 5001.0003.0007, irq 10
+        p11 = re.compile(r'^\s*(?P<intf_number>\d+): '
+                         r'Ext: (?P<interface>[\w\/\.\-]+)\s*: address is '
+                         r'(?P<mac_addr>[0-9A-Fa-f]{4}\.[0-9A-Fa-f]{4}\.[0-9A-Fa-f]{4}), '
```
Comment (LukasMcClelland) [inline src/genie/libs/parser/asa/show_version.py:164]: If you'd prefer, `(?P<mac_addr>[0-9A-Fa-f]{4}\.[0-9A-Fa-f]{4}\.[0-9A-Fa-f]{4})` can be shortened to something like `(?P<mac_addr>[\w\.]+)`. Not a required change

**Review by LukasMcClelland — COMMENTED** [review]
Thank you very much for contributing! Really good code and formatting too!

I left some comments. All except one are optional changes. To implement them or not is entirely up to you. Let me know if you're interested!

**Comment by LukasMcClelland** [comment]
Ah and I see that there's a merge conflict with `sdk_generator/outputs/github_parser.json`

Merging your code with the latest dev branch from https://github.com/CiscoTestAutomation/genieparser is one way to resolve the conflict.

Let me know if you get stuck

**Comment by LukasMcClelland** [comment]
Hello again @ajschroeder. We're looking to reduce the number of open PRs we have currently. 

Do you plan on completing this PR soon? If not, that's not a problem, we can fix it up and merge it for you :)

**Comment by ajschroeder** [comment]
Hi Lukas. I have been caught up in other projects lately, but I would like to finish this and make the changes as you suggested. I do have a little free time today and I can work on adding the license expiry.

AJ

From: Lukas McClelland ***@***.***>
Date: Thursday, July 29, 2021 at 10:37 AM
To: CiscoTestAutomation/genieparser ***@***.***>
Cc: TheHitman1977 ***@***.***>, Mention ***@***.***>
Subject: Re: [CiscoTestAutomation/genieparser] Add ASA ShowVersion (#468)

Hello again @ajschroeder<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fajschroeder&data=04%7C01%7C%7C5c56b1c4344b45983b4608d952a6beaa%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637631698382917441%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=2pXAvd%2B%2F5Hznn1PeLXKu%2FF8a3dPRhxKTw%2BKRAxB5KbI%3D&reserved=0>. We're looking to reduce the number of open PRs we have currently.

Do you plan on completing this PR soon? If not, that's not a problem, we can fix it up and merge it for you :)

**Comment by ajschroeder** [comment]
Hi Lukas,

I have captured the output of the number of remaining days on the licensed feature and also made the changes that you suggested to the regex search patterns. I the tests once again pass so things are looking good from my end. I pushed the new changes up to git for your review.

From: Lukas McClelland ***@***.***>
Date: Thursday, July 29, 2021 at 10:37 AM
To: CiscoTestAutomation/genieparser ***@***.***>
Cc: TheHitman1977 ***@***.***>, Mention ***@***.***>
Subject: Re: [CiscoTestAutomation/genieparser] Add ASA ShowVersion (#468)

Hello again @ajschroeder<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fajschroeder&data=04%7C01%7C%7C5c56b1c4344b45983b4608d952a6beaa%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637631698382917441%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=2pXAvd%2B%2F5Hznn1PeLXKu%2FF8a3dPRhxKTw%2BKRAxB5KbI%3D&reserved=0>. We're looking to reduce the number of open PRs we have currently.

Do you plan on completing this PR soon? If not, that's not a problem, we can fix it up and merge it for you :)

**Comment by LukasMcClelland** [comment]
Excellent, thank you for making the requested changes! Please fix the merge conflict so that some automated tests can run and I'll approve and find a second reviewer

**Comment by ajschroeder** [comment]
Since I’m a git noob, how would I go about fixing this merge conflict? I’m assuming that I need to download the conflicting file and re-push, am I on the right track?

From: Lukas McClelland ***@***.***>
Date: Wednesday, August 4, 2021 at 9:43 AM
To: CiscoTestAutomation/genieparser ***@***.***>
Cc: TheHitman1977 ***@***.***>, Mention ***@***.***>
Subject: Re: [CiscoTestAutomation/genieparser] Add ASA ShowVersion (#468)

Excellent, thank you for making the requested changes! Please fix the merge conflict so that some automated tests can run and I'll approve and find a second reviewer

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCiscoTestAutomation%2Fgenieparser%2Fpull%2F468%23issuecomment-892718260&data=04%7C01%7C%7Ce36f7857cab14111373d08d9575641d1%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637636850253260951%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=IHIzuT8Q7iyEHi3GXJzmqsmkj1GCsqmQpSagPovTwsQ%3D&reserved=0>, or unsubscribe<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FABRCLFRBMGMW2GCMLSNLW33T3FGZ5ANCNFSM46Q2BCOA&data=04%7C01%7C%7Ce36f7857cab14111373d08d9575641d1%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637636850253260951%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=aQbL1obsuF6nP2yInJV34%2BXOj897cEeosc7fHNjrGS8%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7C%7Ce36f7857cab14111373d08d9575641d1%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637636850253270909%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=N%2FA1j9Nv%2BBKAZSmboyd96Ib4dvrC4KTh1lsOWuRfY%2FQ%3D&reserved=0> or Android<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26utm_campaign%3Dnotification-email&data=04%7C01%7C%7Ce36f7857cab14111373d08d9575641d1%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637636850253270909%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=fuZBOlkZGjNPQvSQIAGrGYtl3fF0yJGYSkPnRlw7zZY%3D&reserved=0>.

**Comment by LukasMcClelland** [comment]
Fixing a merge conflict on a fork can be kind of tricky the first few times you do it.

I can fix it for you this time if you'd like. I'll share the commands I use to fix it so that you can fix it yourself next time. Sound alright?

**Comment by ajschroeder** [comment]
Sounds good to me!

**Comment by LukasMcClelland** [comment]
Commands used:
```
git remote add upstream https://github.com/CiscoTestAutomation/genieparser
git fetch upstream 
git merge upstream/master
```

**Review by LukasMcClelland — APPROVED** [review]
Approved! Thank you very much for contributing!

File: src/genie/libs/parser/asa/show_version.py [hunk 4]
```diff
@@ -0,0 +1,856 @@
...
+            line = line.strip()
+
+            # Cisco Adaptive Security Appliance Software Version 9.8(4)10
+            m = p0.match(line)
+            if m:
+                if 'version' not in version_dict:
+                    version_dict['version'] = {}
+                asa_version = m.groupdict()['asa_version']
+                # 9.8(4)10
+                version_dict['version']['asa_version'] = asa_version
```
Comment (domachad) [inline src/genie/libs/parser/asa/show_version.py:284]: No need to change your code, but I think it's worth mentioning: If you use the `setdefault` method, you can make your code cleaner.
For example:
`
version = version_dict.setdefault('version',  {})
`
Once you have `version`, you can do the same for all the other keys, for example:
`
interfaces_dict = version.setdefault('interfaces', {})
`

**Review by domachad — APPROVED** [review]
Great job! Thank you for your contribution! 
I only left a comment that may help you in future contributions.

**Pull request merged**